### ISSUE TITLE
Kubectl Versions >= 1.30 Don't Allow Exec When HTTPS Scheme in proxy-url

### ISSUE THREAD

**Issue by adrianosela** [kubernetes/kubernetes#126134]
### What happened?

Starting on version 1.30, `kubectl exec` craps out when `proxy-url` setting has an HTTPS scheme.

> Note that other operations (e.g. `kubectl get <resource>`) work as expected, it is only `exec` that does not work.

```
10:47 $ cat ~/.kube/config | yq .clusters[0].cluster.proxy-url
https://kind-cluster-from-laptop-olympus.border0.io:1443
```

```
10:48 $ kubectl version
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
```

```
10:49 $ kubectl exec -it alpine-6b45654984-8lqcq -- sh
error: proxy: unknown scheme: https
✘-1 ~
```

Here's the full debug logs:

<details>

```
11:05 $ kubectl --v=8 exec -it alpine-6b45654984-8lqcq -- sh
I0716 11:05:53.288983   62338 loader.go:395] Config loaded from file:  /Users/adriano/.kube/config
I0716 11:05:53.291635   62338 round_trippers.go:463] GET https://kind-cluster-from-laptop-olympus.border0.io/api/v1/namespaces/default/pods/alpine-6b45654984-8lqcq
I0716 11:05:53.291782   62338 round_trippers.go:469] Request Headers:
I0716 11:05:53.291794   62338 round_trippers.go:473]     Accept: application/json, */*
I0716 11:05:53.291797   62338 round_trippers.go:473]     User-Agent: kubectl/v1.30.2 (darwin/arm64) kubernetes/3968350
I0716 11:05:53.730420   62338 round_trippers.go:574] Response Status: 200 OK in 438 milliseconds
I0716 11:05:53.730511   62338 round_trippers.go:577] Response Headers:
I0716 11:05:53.730537   62338 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 0c04f4cf-c63a-4fc5-8f57-3b7c18f9800c
I0716 11:05:53.730553   62338 round_trippers.go:580]     Audit-Id: cf974e60-4df6-4c09-bc44-604a2ff6e9ce
I0716 11:05:53.730564   62338 round_trippers.go:580]     Cache-Control: no-cache, private
I0716 11:05:53.730571   62338 round_trippers.go:580]     Content-Type: application/json
I0716 11:05:53.730579   62338 round_trippers.go:580]     Date: Tue, 16 Jul 2024 18:05:53 GMT
I0716 11:05:53.730586   62338 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: dac17d0d-1c79-4632-a567-77ac067e87a1
I0716 11:05:53.731323   62338 request.go:1212] Response Body: {"kind":"Pod","apiVersion":"v1","metadata":{"name":"alpine-6b45654984-8lqcq","generateName":"alpine-6b45654984-","namespace":"default","uid":"c852cc5b-7a05-4221-a165-de67020a1e21","resourceVersion":"554","creationTimestamp":"2024-07-16T00:13:37Z","labels":{"app":"alpine","pod-template-hash":"6b45654984"},"ownerReferences":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"alpine-6b45654984","uid":"003afd22-0098-4ed8-bdee-0490f4f6533c","controller":true,"blockOwnerDeletion":true}],"managedFields":[{"manager":"kube-controller-manager","operation":"Update","apiVersion":"v1","time":"2024-07-16T00:13:37Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:app":{},"f:pod-template-hash":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"003afd22-0098-4ed8-bdee-0490f4f6533c\"}":{}}},"f:spec":{"f:containers":{"k:{\"name\":\"alpine\"}":{".":{},"f:command":{},"f:image":{},"f:imagePullPolicy":{},"f:name":{},"f:resources":{},"f:terminationMessagePath":{},"f:terminationMessagePolicy":{ [truncated 3749 chars]
I0716 11:05:53.737452   62338 podcmd.go:88] Defaulting container name to alpine
I0716 11:05:53.738803   62338 round_trippers.go:463] GET https://kind-cluster-from-laptop-olympus.border0.io/api/v1/namespaces/default/pods/alpine-6b45654984-8lqcq/exec?command=sh&container=alpine&stdin=true&stdout=true&tty=true
                I0716 11:05:53.738828   62338 round_trippers.go:469] Request Headers:
                                                                                     I0716 11:05:53.738840   62338 round_trippers.go:473]     Sec-Websocket-Protocol: v5.channel.k8s.io
                                                                             I0716 11:05:53.738849   62338 round_trippers.go:473]     User-Agent: kubectl/v1.30.2 (darwin/arm64) kubernetes/3968350
                                                                                         I0716 11:05:53.738891   62338 round_trippers.go:574] Response Status:  in 0 milliseconds
                                                                       I0716 11:05:53.738901   62338 round_trippers.go:577] Response Headers:
                                   error: proxy: unknown scheme: https
```

</details>

This does not happen if I downgrade my client to v 1.29

```
10:57 $ kubectl version
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
```

```
10:58 $ kubectl exec -it alpine-6b45654984-8lqcq -- sh
/ # ls
bin           home          mnt           product_uuid  sbin          tmp
dev           lib           opt           root          srv           usr
etc           media         proc          run           sys           var
/ # exit
✔ ~
```

### What did you expect to happen?

I expect `kubectl` to allow and respect `proxy-url`s  with HTTPS scheme.

### How can we reproduce it (as minimally and precisely as possible)?

See "what happened" section. Set up an HTTPS proxy, and note that v1.30+ do not work.

### Anything else we need to know?

N.A

### Kubernetes version

<details>

```console
10:57 $ kubectl version
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
```

</details>

### Cloud provider

<details>

N/A - Kind cluster on my Macbook

</details>

### OS version

<details>

N/A - Kind cluster on my Macbook

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by BenTheElder** [comment]
/sig cli
cc @seans3

**Comment by adrianosela** [comment]
Here's some more info on the behaviour diff between v1.29 and v1.30.

Both versions have the same initial request to get pod info:

```
GET /api/v1/namespaces/${NAMESPACE}/pods/${POD}
```

However the second request is different for v1.29 and v1.30:

#### 1.29:

```
POST /api/v1/namespaces/${NAMESPACE}/pods/${POD}/exec?command=${COMMAND}&container=${CONTAINER}&stdin=true&stdout=true&tty=true
```

#### 1.30:

```
GET /api/v1/namespaces/${NAMESPACE}/pods/${POD}/exec?command=${COMMAND}&container=${CONTAINER}&stdin=true&stdout=true&tty=true
```

---

Logs from (successful) exec with v1.29:

<details>

```
✔ ~
12:08 $ kubectl version
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
✔ ~
12:08 $ kubectl --v=8 exec -it alpine-6b45654984-8lqcq -- sh
I0716 12:08:39.539199   67029 loader.go:395] Config loaded from file:  /Users/adriano/.kube/config
I0716 12:08:39.543538   67029 round_trippers.go:463] GET https://kind-cluster-from-laptop-olympus.border0.io/api/v1/namespaces/default/pods/alpine-6b45654984-8lqcq
I0716 12:08:39.543554   67029 round_trippers.go:469] Request Headers:
I0716 12:08:39.543558   67029 round_trippers.go:473]     Accept: application/json, */*
I0716 12:08:39.543563   67029 round_trippers.go:473]     User-Agent: kubectl/v1.29.2 (darwin/arm64) kubernetes/4b8e819
I0716 12:08:39.957884   67029 round_trippers.go:574] Response Status: 200 OK in 414 milliseconds
I0716 12:08:39.957952   67029 round_trippers.go:577] Response Headers:
I0716 12:08:39.957974   67029 round_trippers.go:580]     Cache-Control: no-cache, private
I0716 12:08:39.957991   67029 round_trippers.go:580]     Content-Type: application/json
I0716 12:08:39.958008   67029 round_trippers.go:580]     Date: Tue, 16 Jul 2024 19:08:39 GMT
I0716 12:08:39.958024   67029 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: dac17d0d-1c79-4632-a567-77ac067e87a1
I0716 12:08:39.958041   67029 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 0c04f4cf-c63a-4fc5-8f57-3b7c18f9800c
I0716 12:08:39.958058   67029 round_trippers.go:580]     Audit-Id: a7f49d43-2fb5-48d8-b742-cd22f0806cd7
I0716 12:08:39.984305   67029 request.go:1212] Response Body: {"kind":"Pod","apiVersion":"v1","metadata":{"name":"alpine-6b45654984-8lqcq","generateName":"alpine-6b45654984-","namespace":"default","uid":"c852cc5b-7a05-4221-a165-de67020a1e21","resourceVersion":"554","creationTimestamp":"2024-07-16T00:13:37Z","labels":{"app":"alpine","pod-template-hash":"6b45654984"},"ownerReferences":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"alpine-6b45654984","uid":"003afd22-0098-4ed8-bdee-0490f4f6533c","controller":true,"blockOwnerDeletion":true}],"managedFields":[{"manager":"kube-controller-manager","operation":"Update","apiVersion":"v1","time":"2024-07-16T00:13:37Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:app":{},"f:pod-template-hash":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"003afd22-0098-4ed8-bdee-0490f4f6533c\"}":{}}},"f:spec":{"f:containers":{"k:{\"name\":\"alpine\"}":{".":{},"f:command":{},"f:image":{},"f:imagePullPolicy":{},"f:name":{},"f:resources":{},"f:terminationMessagePath":{},"f:terminationMessagePolicy":{ [truncated 3749 chars]
I0716 12:08:39.986850   67029 podcmd.go:88] Defaulting container name to alpine
I0716 12:08:39.988199   67029 round_trippers.go:463] POST https://kind-cluster-from-laptop-olympus.border0.io/api/v1/namespaces/default/pods/alpine-6b45654984-8lqcq/exec?command=sh&container=alpine&stdin=true&stdout=true&tty=true
I0716 12:08:39.988212   67029 round_trippers.go:469] Request Headers:
I0716 12:08:39.988219   67029 round_trippers.go:473]     User-Agent: kubectl/v1.29.2 (darwin/arm64) kubernetes/4b8e819
I0716 12:08:39.988225   67029 round_trippers.go:473]     X-Stream-Protocol-Version: v5.channel.k8s.io
I0716 12:08:39.988232   67029 round_trippers.go:473]     X-Stream-Protocol-Version: v4.channel.k8s.io
I0716 12:08:39.988237   67029 round_trippers.go:473]     X-Stream-Protocol-Version: v3.channel.k8s.io
I0716 12:08:39.988243   67029 round_trippers.go:473]     X-Stream-Protocol-Version: v2.channel.k8s.io
I0716 12:08:39.988248   67029 round_trippers.go:473]     X-Stream-Protocol-Version: channel.k8s.io
I0716 12:08:40.391675   67029 round_trippers.go:574] Response Status: 101 Switching Protocols in 403 milliseconds
I0716 12:08:40.391721   67029 round_trippers.go:577] Response Headers:
I0716 12:08:40.391733   67029 round_trippers.go:580]     Date: Tue, 16 Jul 2024 19:08:40 GMT
I0716 12:08:40.391743   67029 round_trippers.go:580]     Upgrade: SPDY/3.1
I0716 12:08:40.391791   67029 round_trippers.go:580]     X-Stream-Protocol-Version: v4.channel.k8s.io
I0716 12:08:40.391803   67029 round_trippers.go:580]     Content-Length: 0
I0716 12:08:40.391812   67029 round_trippers.go:580]     Connection: Upgrade
/ # ls
bin           etc           lib           mnt           proc          root          sbin          sys           usr
dev           home          media         opt           product_uuid  run           srv           tmp           var
/ # exit
```

</details>


Logs from (successful) exec with v1.30:

<details>

```
✔ ~
12:05 $ kubectl version
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
✔ ~
12:05 $ kubectl --v=8 exec -it alpine-6b45654984-8lqcq -- sh
I0716 12:10:41.458375   67178 loader.go:395] Config loaded from file:  /Users/adriano/.kube/config
I0716 12:10:41.468315   67178 round_trippers.go:463] GET https://kind-cluster-from-laptop-olympus.border0.io/api/v1/namespaces/default/pods/alpine-6b45654984-8lqcq
I0716 12:10:41.468327   67178 round_trippers.go:469] Request Headers:
I0716 12:10:41.468332   67178 round_trippers.go:473]     Accept: application/json, */*
I0716 12:10:41.468334   67178 round_trippers.go:473]     User-Agent: kubectl/v1.30.2 (darwin/arm64) kubernetes/3968350
I0716 12:10:41.904233   67178 round_trippers.go:574] Response Status: 200 OK in 435 milliseconds
I0716 12:10:41.904272   67178 round_trippers.go:577] Response Headers:
I0716 12:10:41.904300   67178 round_trippers.go:580]     Audit-Id: 393af37f-0a8e-458c-9cc4-2501165572e5
I0716 12:10:41.904308   67178 round_trippers.go:580]     Cache-Control: no-cache, private
I0716 12:10:41.904314   67178 round_trippers.go:580]     Content-Type: application/json
I0716 12:10:41.904319   67178 round_trippers.go:580]     Date: Tue, 16 Jul 2024 19:10:41 GMT
I0716 12:10:41.904324   67178 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: dac17d0d-1c79-4632-a567-77ac067e87a1
I0716 12:10:41.904329   67178 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 0c04f4cf-c63a-4fc5-8f57-3b7c18f9800c
I0716 12:10:41.905625   67178 request.go:1212] Response Body: {"kind":"Pod","apiVersion":"v1","metadata":{"name":"alpine-6b45654984-8lqcq","generateName":"alpine-6b45654984-","namespace":"default","uid":"c852cc5b-7a05-4221-a165-de67020a1e21","resourceVersion":"554","creationTimestamp":"2024-07-16T00:13:37Z","labels":{"app":"alpine","pod-template-hash":"6b45654984"},"ownerReferences":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"alpine-6b45654984","uid":"003afd22-0098-4ed8-bdee-0490f4f6533c","controller":true,"blockOwnerDeletion":true}],"managedFields":[{"manager":"kube-controller-manager","operation":"Update","apiVersion":"v1","time":"2024-07-16T00:13:37Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:app":{},"f:pod-template-hash":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"003afd22-0098-4ed8-bdee-0490f4f6533c\"}":{}}},"f:spec":{"f:containers":{"k:{\"name\":\"alpine\"}":{".":{},"f:command":{},"f:image":{},"f:imagePullPolicy":{},"f:name":{},"f:resources":{},"f:terminationMessagePath":{},"f:terminationMessagePolicy":{ [truncated 3749 chars]
I0716 12:10:41.910863   67178 podcmd.go:88] Defaulting container name to alpine
I0716 12:10:41.911911   67178 round_trippers.go:463] GET https://kind-cluster-from-laptop-olympus.border0.io/api/v1/namespaces/default/pods/alpine-6b45654984-8lqcq/exec?command=sh&container=alpine&stdin=true&stdout=true&tty=true
I0716 12:10:41.911932   67178 round_trippers.go:469] Request Headers:
I0716 12:10:41.911940   67178 round_trippers.go:473]     Sec-Websocket-Protocol: v5.channel.k8s.io
I0716 12:10:41.911944   67178 round_trippers.go:473]     User-Agent: kubectl/v1.30.2 (darwin/arm64) kubernetes/3968350
I0716 12:10:41.911970   67178 round_trippers.go:574] Response Status:  in 0 milliseconds
I0716 12:10:41.911977   67178 round_trippers.go:577] Response Headers:
                                                                                                                                                                    error: proxy: unknown scheme: https
```

</details>

**Comment by adrianosela** [comment]
Update: Setting `KUBECTL_REMOTE_COMMAND_WEBSOCKETS=false` makes the error go away.

```
✔ ~
12:26 $ KUBECTL_REMOTE_COMMAND_WEBSOCKETS=false kubectl version
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
✔ ~
12:26 $ KUBECTL_REMOTE_COMMAND_WEBSOCKETS=false kubectl exec -it alpine-6b45654984-8lqcq -- sh
/ # ls
bin           etc           lib           mnt           proc          root          sbin          sys           usr
dev           home          media         opt           product_uuid  run           srv           tmp           var
/ # exit
```

**Comment by dims** [comment]
cc @seans3

**Comment by adrianosela** [comment]
I added a bunch of print statements and built `kubectl` from source (`make kubectl` in this repo)...

I found that the error:

```
proxy: unknown scheme: https
```

Is returned by `dialer.DialContext()` in the `RoundTrip(...)` method of the `RoundTripper` iface implementation in `/staging/src/k8s.io/client-go/transport/websocket/roundtripper.go`

**Comment by adrianosela** [comment]
Long-lived open issue in the gorilla/websockets repo: https://github.com/gorilla/websocket/issues/794

And a **closed** PR which attempted to address the lack of HTTPS proxy support: https://github.com/gorilla/websocket/pull/916

**Comment by adrianosela** [comment]
**TL;DR;**

Changing the default behaviour of feature gate `KUBECTL_REMOTE_COMMAND_WEBSOCKETS` from disabled to enabled in `kubectl` release version 1.30 breaks support for `kubectl exec` over HTTPS proxy servers. :)

**Comment by ah8ad3** [comment]
I am looking into it, if @seans3 is not available i will give it a try.

**Comment by adrianosela** [comment]
Support is being added in https://github.com/gorilla/websocket/pull/950

**Comment by seans3** [comment]
@adrianosela Thanks for all the useful information on this issue, and for help in its resolution. Would it be possible to enumerate some specific instructions on how reproduce the issue, including how to set up a specific HTTPS proxy? This will also help write tests to validate the solution and ensure no regressions. Thanks :)

More context: This issue is caused by transitioning the k8s streaming protocol from SPDY to WebSockets (https://github.com/kubernetes/enhancements/issues/4006). The transition to WebSockets moved to Beta in v1.30, and it became the default for the `RemoteCommand` subprotocol (e.g. `kubectl exec`, `kubectl attach`, and `kubectl cp`).  This issue is also present in v1.29, but it is hidden behind feature flags (KUBECTL_REMOTE_COMMAND_WEBSOCKETS) which are off by default. The `PortForward` subprotocol (`kubectl port-forward`) will become the default in the upcoming v1.31 release. This issue will also affect the `PortForward` efforts.

**Comment by adrianosela** [comment]
Hey @seans3 

# How to reproduce the issue

I ran into the issue while testing my company's product (identity/application aware proxies) beta support for Kubernetes. Trying that out is probably the easiest/fastest way for you to reproduce the issue. (I swear I'm not trying to promote it here lol).

It's free for personal use cases and really fast to get going. Here's a video I made for a customer that might be helpful to you if you do give it a shot: https://drive.google.com/file/d/1I0xmkJ-4ybuzObABt69e11O8m-NVwF4F/view

`kubectl` < v1.30 should work without any changes, `kubectl` >= v1.30 should only work with the env `KUBECTL_REMOTE_COMMAND_WEBSOCKETS=false`

I assume Teleport and other "proxy as a service" services out there also use HTTPS proxies for Kubernets support as the proxies are authenticated. (are people still really using basic-authenticated proxies over HTTP in 2024? ...yes they are...)

# How to set up an HTTPS proxy in go:

On a high level to set up an HTTPS proxy in go:

- generate a self-signed cert / key-pair to use for TLS
  - [mkcert](https://github.com/FiloSottile/mkcert) automates adding the self-signed CA to your machine's trust-store
- set up an `http.Server` object and call `ListenAndServeTLS(/* args */)` on it

<details>

<summary>pseudocode</summary>

```
server := &http.Server{
    Addr:  "localhost:8443",
    Handler: handler,
    TLSConfig: &tls.Config{
	// Must force HTTP/1.1 to be used for HTTP CONNECT based sockets
        // since HTTP/2.0 does not support the CONNECT HTTP method.
        NextProtos = []string{"http/1.1"}
    },
}

if err := server.ListenAndServeTLS("/path/to/your/cert.pem", "/path/to/your/key.pem"); err != nil {
    return fmt.Errorf("failed to listen and serve tls: %v", err)
}
```

</details>

- the handler in the server above (`http.Handler`) should handle the HTTP CONNECT method and set up a full duplex connection between the client and the remote server

<details>

<summary>pseudocode</summary>

```
handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
     if r.Method != http.MethodConnect {
        http.Error(w, fmt.Sprintf("only %s is allowed, but got %s", http.MethodConnect, r.Method), http.StatusMethodNotAllowed)
        return
    }

    // NOTE: parse the target address out of the CONNECT request... or just hardcode it
    tunnelToDestination(targetAddress, cfg.TunnelDialTimeout, w)
})

// tunnelToDestination connects to a remote server and sets up a full
// duplex connection between the client and the remote server
func tunnelToDestination(
	targetAddress string,
	dialTimeout time.Duration,
	w http.ResponseWriter,
) {
        // if the destination is HTTPS then use tls.Dial(/* args */) here, or wrap the tcp net.Conn in a tls.Client
	toDestination, err := net.DialTimeout("tcp", targetAddress, dialTimeout)
	if err != nil {
                 // probably log or something
		http.Error(w, "an unknown error occurred... try again later", http.StatusServiceUnavailable)
		return
	}

	w.WriteHeader(http.StatusOK)

	hijacker, ok := w.(http.Hijacker)
	if !ok {
                 // probably log or something
		http.Error(w, "an unknown error occurred... try again later", http.StatusInternalServerError)
		return
	}

	toClient, _, err := hijacker.Hijack()
	if err != nil {
                 // probably log or something
		http.Error(w, "an unknown error occurred... try again later", http.StatusServiceUnavailable)
		return
	}

	if err = iocopy(toClient, toDestination); err != nil {
                 // probably log or something
		http.Error(w, "an unknown error occurred... try again later", http.StatusServiceUnavailable)
		return
	}
}

func iocopy(src, dst net.Conn) error {
	defer src.Close()
	defer dst.Close()

	errC := make(chan error, 2)
	defer close(errC)

	var wg sync.WaitGroup
	defer wg.Wait()

	wg.Add(1)
	go func() {
		defer wg.Done()

		// abort operations on dst when src has an error (incl. EOF)
		defer dst.SetDeadline(time.Now().Add(-1 * time.Hour))

		_, err := io.Copy(dst, src)
		errC <- err
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()

		// abort operations on src when dst has an error (incl. EOF)
		defer src.SetDeadline(time.Now().Add(-1 * time.Hour))

		_, err := io.Copy(src, dst)
		errC <- err
	}()

	// returns the first error encountered
	return <-errC
}
```

</details>

----

^^ I hope this helps.

I intend to try building `kubectl` with the latest release of `gorilla/websocket` whenever they merge that PR I linked earlier and reporting back here.

I am also happy to help write the regression tests afterwards.

**Comment by adrianosela** [comment]
Hey @seans3, see the README/code here https://github.com/adrianosela/https-proxy. Hopefully this will save you tons of time. I just cleaned up the code from my previous reply and added some notes on how to fully reproduce the issue.
